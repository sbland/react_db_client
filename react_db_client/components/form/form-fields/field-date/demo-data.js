export const defaultVal = new Date(Date.now());