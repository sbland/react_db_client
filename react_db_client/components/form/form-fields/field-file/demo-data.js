export const DEMO_FILES_DATA = [
    { uid: 'a', filePath: 'files/fileA.jpg', name: 'FileA' },
    { uid: 'b', filePath: 'files/fileB.jpg', name: 'FileA' },
    { uid: 'c', filePath: 'files/fileC.jpg', name: 'FileA' },
  ];
