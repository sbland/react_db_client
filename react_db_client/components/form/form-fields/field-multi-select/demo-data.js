export const demoOptions = [
    {
        uid: 'a',
        label: 'A',
    },
    {
        uid: 'b',
        label: 'B',
    },
    {
        uid: 'c',
        label: 'C',
    },
];