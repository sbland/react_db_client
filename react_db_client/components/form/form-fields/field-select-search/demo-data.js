export const defaultVal = 'a';

export const demoOptions = [
    {
        uid: 'a',
        label: 'a',
    },{
        uid: 'b',
        label: 'b',
    },{
        uid: 'c',
        label: 'c',
    }
];
