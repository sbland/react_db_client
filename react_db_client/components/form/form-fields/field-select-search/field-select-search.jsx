import React, { useCallback, useMemo } from 'react';
import PropTypes from 'prop-types';
import { SearchAndSelectDropdown } from '@react_db_client/components.search-and-select-dropdown';

export const ShowMultiSelection = ({ values, labelField, onSelect }) => {
  return (
    <div>
      {values?.map((v) => (
        <button onClick={() => onSelect(v)} key={v.uid || v}>
          {v[labelField] || v.uid || v}
        </button>
      ))}
    </div>
  );
};

export const FieldSelectSearch = ({
  uid,
  unit,
  updateFormData,
  value,
  multiple,
  required,
  searchFn,
  returnFieldOnSelect, // the field in the data to return
  searchFieldTargetField, // the target field that the search string applies to
  labelField, // The field in the returned data to use as the label
  allowEmptySearch,
  className,
}) => {
  const valueIn = useMemo(() => (multiple ? '' : value), [multiple, value]);

  const handleSelect = useCallback(
    (_, data) => {
      let newVal = returnFieldOnSelect ? data[returnFieldOnSelect] : data;
      if (multiple) {
        if (Array.isArray(value)) {
          const newValueArray = value.concat(data);
          newVal = newValueArray;
        } else {
          newVal = [data];
        }
      }
      updateFormData(uid, newVal);
    },
    [multiple, updateFormData, returnFieldOnSelect]
  );

  const clickSelectedItem = useCallback((selectedItem) => {
    if (Array.isArray(value)) {
      const newValueArray = value.filter((v) => (v.uid != null || v.uid !== selectedItem.uid) && v !== selectedItem);
      updateFormData(uid, newValueArray);
    } else {
      updateFormData(uid, []);
    }
  }, []);

  return (
    <>
      <SearchAndSelectDropdown
        searchFunction={searchFn}
        initialValue={valueIn}
        handleSelect={handleSelect}
        searchFieldTargetField={searchFieldTargetField}
        labelField={labelField}
        className={className}
        allowEmptySearch={allowEmptySearch}
        searchFieldPlaceholder={`${value || 'search...'}`}
        required={required}
      />
      {multiple && (
        <ShowMultiSelection onSelect={clickSelectedItem} labelField={labelField} values={value} />
      )}
    </>
  );
};

FieldSelectSearch.propTypes = {
  uid: PropTypes.string.isRequired,
  unit: PropTypes.string,
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
    PropTypes.shape({ uid: PropTypes.string }),
    PropTypes.arrayOf(
      PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.number,
        PropTypes.shape({ uid: PropTypes.string }),
      ])
    ),
  ]),
  // defaultVal: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  updateFormData: PropTypes.func.isRequired,
  multiple: PropTypes.bool,
  required: PropTypes.bool,
  searchFn: PropTypes.func.isRequired,
  returnFieldOnSelect: PropTypes.string,
  searchFieldTargetField: PropTypes.string.isRequired,
  labelField: PropTypes.string,
  allowEmptySearch: PropTypes.bool,
  className: PropTypes.string,
};

FieldSelectSearch.defaultProps = {
  unit: '',
  value: null,
  multiple: false,
  // defaultVal: null,
  required: false,
  returnFieldOnSelect: 'uid',
  labelField: 'label',
  allowEmptySearch: false,
  className: '',
};
