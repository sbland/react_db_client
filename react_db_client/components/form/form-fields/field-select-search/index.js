export { FieldSelectSearch } from './field-select-search';
