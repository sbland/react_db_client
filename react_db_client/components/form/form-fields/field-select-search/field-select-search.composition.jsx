import React from 'react';
import { CompositionWrapDefault } from '@samnbuk/react_db_client.helpers.composition-wraps';
import { FieldSelectSearch } from './field-select-search';
import { defaultVal, demoOptions } from './demo-data';

const updateFormData = () => {};
const searchFn = async () => demoOptions;

const defaultProps = {
  uid: 'demoid',
  unit: 'demounit',
  updateFormData,
  value: defaultVal,
  multiple: false,
  required: false,
  searchFn,
  returnFieldOnSelect: 'uid',
  searchFieldTargetField: 'label',
  labelField: 'label',
};

export const BasicFieldSelectSearch = () => (
  <CompositionWrapDefault height="4rem" width="8rem">
    <FieldSelectSearch {...defaultProps} required />
  </CompositionWrapDefault>
);
