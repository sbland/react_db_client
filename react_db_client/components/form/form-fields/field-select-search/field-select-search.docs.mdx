---
description: 'A React Form selection field with search built in.'
labels: ['form-field', 'field-select-search']
---

import { FieldSelectSearch } from './field-select-search';

