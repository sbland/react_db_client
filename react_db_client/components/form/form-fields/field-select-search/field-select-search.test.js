import '@samnbuk/react_db_client.testing.enzyme-setup';
import React from 'react';
import { mount, shallow } from 'enzyme';
import { act } from 'react-dom/test-utils';
import { MockReactC } from '@react_db_client/testing.utils';

import { SearchAndSelectDropdown } from '@react_db_client/components.search-and-select-dropdown';

import { FieldSelectSearch, ShowMultiSelection } from './field-select-search';
import * as compositions from './field-select-search.composition';
import { defaultVal } from './demo-data';

jest.mock('@react_db_client/components.search-and-select-dropdown', () =>
  MockReactC('SearchAndSelectDropdown', ['SearchAndSelectDropdown'])
);

const updateFormData = jest.fn();
const searchFn = jest.fn();

const defaultProps = {
  uid: 'demoid',
  unit: 'demounit',
  updateFormData,
  value: defaultVal,
  multiple: false,
  required: false,
  searchFn,
  returnFieldOnSelect: 'uid',
  searchFieldTargetField: 'name',
  labelField: 'name',
};

describe('field-select-search', () => {
  beforeEach(() => {
    updateFormData.mockClear();
    searchFn.mockClear();
  });
  test('Renders', () => {
    shallow(<FieldSelectSearch {...defaultProps} />);
  });

  describe('Compositions', () => {
    Object.entries(compositions).forEach(([name, Composition]) => {
      test(name, () => {
        mount(<Composition />);
      });
    });
  });
  describe('Snapshots', () => {
    test('Matches Snapshot', () => {
      const component = mount(<FieldSelectSearch {...defaultProps} />);
      const tree = component.debug();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('Unit Tests', () => {
    describe('Default', () => {
      let component;
      beforeEach(() => {
        component = mount(<FieldSelectSearch {...defaultProps} />);
      });
      describe('Searching', () => {
        /* Searching handled by search and select dropdown. */
        test('should call search function', () => {
          const searchAndSelectDropdown = component.find(SearchAndSelectDropdown);
          const searchText = 'searchText';
          act(() => {
            searchAndSelectDropdown.props().searchFunction(searchText);
          });
          expect(searchFn).toHaveBeenCalledWith(searchText);
        });
      });
      describe('shows a search select field', () => {
        test('should show search and select component', () => {
          expect(component.find(SearchAndSelectDropdown)).toBeTruthy();
        });
      });
      describe('handles select', () => {
        test('should call update fn when search component returns selection', () => {
          const searchComponent = component.find(SearchAndSelectDropdown);
          const selectedId = 'demoid';
          const selectedData = { uid: selectedId };
          searchComponent.props().handleSelect(selectedId, selectedData);
          expect(updateFormData).toHaveBeenCalledWith(defaultProps.uid, selectedId);
        });
      });
    });

    describe('Multiple', () => {
      const firstSelection = { uid: 'demoid1', label: 'Demo Label 1' };
      const secondSelection = { uid: 'demoid2', label: 'Demo Label 2' };

      test('Display search field', () => {
        const component = mount(<FieldSelectSearch {...defaultProps} multiple value={null} />);
        const searchComponent = component.find(SearchAndSelectDropdown);
        searchComponent.props().handleSelect(firstSelection.uid, firstSelection);
        expect(updateFormData).toHaveBeenCalledWith(defaultProps.uid, [firstSelection]);
      });

      test('Display current selection labels', () => {
        const component = mount(
          <FieldSelectSearch {...defaultProps} multiple value={[firstSelection]} />
        );
        const showMultiSelection = component.find(ShowMultiSelection);
        expect(showMultiSelection.props().values).toEqual([firstSelection]);
      });

      test('Can remove from selection', () => {
        const component = mount(
          <FieldSelectSearch {...defaultProps} multiple value={[firstSelection, secondSelection]} />
        );
        const showMultiSelection = component.find(ShowMultiSelection);
        const firstSelectionButton = showMultiSelection.find('button').at(0);
        act(() => {
          firstSelectionButton.simulate('click');
        });
        expect(updateFormData).toHaveBeenCalledWith(defaultProps.uid, [secondSelection]);
      });
    });
  });
});
