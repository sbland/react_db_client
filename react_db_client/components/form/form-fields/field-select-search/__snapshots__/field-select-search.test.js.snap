// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`field-select-search Snapshots Matches Snapshot 1`] = `
"<FieldSelectSearch uid=\\"demoid\\" unit=\\"demounit\\" updateFormData={[Function: mockConstructor] { _isMockFunction: true, getMockImplementation: [Function (anonymous)], mock: Object [Object: null prototype] { calls: [], instances: [], invocationCallOrder: [], results: [] }, mockClear: [Function (anonymous)], mockReset: [Function (anonymous)], mockRestore: [Function (anonymous)], mockReturnValueOnce: [Function (anonymous)], mockResolvedValueOnce: [Function (anonymous)], mockRejectedValueOnce: [Function (anonymous)], mockReturnValue: [Function (anonymous)], mockResolvedValue: [Function (anonymous)], mockRejectedValue: [Function (anonymous)], mockImplementationOnce: [Function (anonymous)], mockImplementation: [Function (anonymous)], mockReturnThis: [Function (anonymous)], mockName: [Function (anonymous)], getMockName: [Function (anonymous)] }} value=\\"a\\" multiple={false} required={false} searchFn={[Function: mockConstructor] { _isMockFunction: true, getMockImplementation: [Function (anonymous)], mock: Object [Object: null prototype] { calls: [], instances: [], invocationCallOrder: [], results: [] }, mockClear: [Function (anonymous)], mockReset: [Function (anonymous)], mockRestore: [Function (anonymous)], mockReturnValueOnce: [Function (anonymous)], mockResolvedValueOnce: [Function (anonymous)], mockRejectedValueOnce: [Function (anonymous)], mockReturnValue: [Function (anonymous)], mockResolvedValue: [Function (anonymous)], mockRejectedValue: [Function (anonymous)], mockImplementationOnce: [Function (anonymous)], mockImplementation: [Function (anonymous)], mockReturnThis: [Function (anonymous)], mockName: [Function (anonymous)], getMockName: [Function (anonymous)] }} returnFieldOnSelect=\\"uid\\" searchFieldTargetField=\\"name\\" labelField=\\"name\\" allowEmptySearch={false} className=\\"\\">
  <SearchAndSelectDropdown searchFunction={[Function: mockConstructor] { _isMockFunction: true, getMockImplementation: [Function (anonymous)], mock: Object [Object: null prototype] { calls: [], instances: [], invocationCallOrder: [], results: [] }, mockClear: [Function (anonymous)], mockReset: [Function (anonymous)], mockRestore: [Function (anonymous)], mockReturnValueOnce: [Function (anonymous)], mockResolvedValueOnce: [Function (anonymous)], mockRejectedValueOnce: [Function (anonymous)], mockReturnValue: [Function (anonymous)], mockResolvedValue: [Function (anonymous)], mockRejectedValue: [Function (anonymous)], mockImplementationOnce: [Function (anonymous)], mockImplementation: [Function (anonymous)], mockReturnThis: [Function (anonymous)], mockName: [Function (anonymous)], getMockName: [Function (anonymous)] }} initialValue=\\"a\\" handleSelect={[Function (anonymous)]} searchFieldTargetField=\\"name\\" labelField=\\"name\\" className=\\"\\" allowEmptySearch={false} searchFieldPlaceholder=\\"a\\" required={false}>
    <div data-testid=\\"SearchAndSelectDropdown\\">
      SearchAndSelectDropdown
    </div>
  </SearchAndSelectDropdown>
</FieldSelectSearch>"
`;
